### ISSUE TITLE
Fix display scaling on linux

### ISSUE THREAD

**Issue by HermanKirshin** [AvaloniaUI/Avalonia#9390]
I have tested how application is displayed in 7 different desktop environments with multiple monitors (combination of high-dpi and low-dpi), some of them cover 90% of cases I think. Here are the results:

Default Avalonia behavior:
- 	**Gnome**
    	Always rendered in 100% 
			**incorrect** (Scale settings does not affect, looks too small on HiDpi displays)
- 	**KDE, XFCE, MATE, LXQt, Deepin**
    	always rendered 200% on HiDpi display and 100% on LowDpi display
			**incorrect**, displays may be configured to 150% and app will be bigger than expected, and on low dpi smaller than expected.
		on virtual machine always rendered 100% because virtual monitor does not report physical size and HiDpi detection logic fails	
- 	**Cinnamon**
		if all monitors set to 100% then rendered 200% on HiDpi monitor and 100% on LoDpi monitor 
			**incorrect**, HiDpi may be configured to 100 or 150 and app will be bigger than expected, and on low dpi smaller than expected.
        Configure HiDpi to 125% and LoDpi to 100% then displayed on HiDpi in 62%  and on LoDpi in 50%   
        Configure HiDpi to 150% and LoDpi to 100% then displayed on HiDpi in 75%  and on LoDpi in 50%   
        Configure HiDpi to 175% and LoDpi to 100% then displayed on HiDpi in 87%  and on LoDpi in 50%   
        Configure HiDpi to 200% and LoDpi to 100% then displayed on HiDpi in 200% and on LoDpi in 50%   
        Configure HiDpi to 100% and LoDpi to 125% then displayed on HiDpi in 50%  and on LoDpi in 62%   
        Configure HiDpi to 100% and LoDpi to 150% then displayed on HiDpi in 50%  and on LoDpi in 75%   
        Configure HiDpi to 100% and LoDpi to 175% then displayed on HiDpi in 50%  and on LoDpi in 82%   
        Configure HiDpi to 100% and LoDpi to 200% then displayed on HiDpi in 50%  and on LoDpi in 100%  
        Configure HiDpi to 150% and LoDpi to 150% then displayed on HiDpi in 75%  and on LoDpi in 75%   
			**incorrect**, app looks even smaller while all other apps are displayed correctly (unlike XFCE where everything looks weird)
	
I have looked into X11Screens code, ant it does very strange things.
		
Conclusion - looks like there is **NO** linux setups where avalonia apps are scaled correctly, while majority of up-to-date apps are displayed as expected respecting scale factors from system settings. 

Missing such fundamental feature is a serious flaw, because most other UI frameworks are capable to do it. If they can, why we cannot? It is possible to get with high probability correct values from system, that will cover majority of cases, here is some prototype:
https://github.com/HermanKirshin/LinuxDisplayScale/blob/main/Program.cs

Currently I use this to setup AVALONIA_SCREEN_SCALE_FACTORS before calling AppBuilder

Also here is some investigation summary about desktop environments https://github.com/HermanKirshin/LinuxDisplayScale/blob/main/linux_scale_info.txt

**Comment by HermanKirshin** [comment]
#4826

**Comment by kekekeks** [comment]
X11 **does not support per display scaling**. Any values you set in your system settings in your Wayland-powered DE **are invisible to X11 apps**. Technically nothing prevents DE authors from providing per-display DPI information to X11 apps, but they deliberately choose not to.

Avalonia currently has the same behavior as Qt5 in automatic DPI mode which is DPI detection based on display EDID with an environment variable override. EDID information is incorrect with XWayland.

TL;DR: Wait for a Wayland backend to be ready or switch to pure X11 session

**Comment by HermanKirshin** [comment]
<img width="1079" alt="Screenshot from 2022-11-09 09-06-54" src="https://user-images.githubusercontent.com/25755704/200764598-6065d710-ab3b-4b9c-b86b-58352fdce79e.png">

**Comment by kekekeks** [comment]
You are running a Wayland session. Your Qt app is running with Wayland support.

**Comment by timunie** [comment]
If you want to try out the Wayland backend, you can find the PR here: #8352 

Any feedback about it would be very helpful I think. 

@kekekeks should we mark this issue as by-design and move it to Q&A?

**Comment by HermanKirshin** [comment]
@kekekeks Why do you think that it is Wayland session?

<img width="1033" alt="Screenshot 2022-11-09 094548" src="https://user-images.githubusercontent.com/25755704/200769532-ca2fd720-d779-44db-a14a-556a2cac50b5.png">

**Comment by timunie** [comment]
@HermanKirshin I think you are right 👍

**Comment by HermanKirshin** [comment]
Made some important refactorings in scale factor detection algorithm https://github.com/HermanKirshin/LinuxDisplayScale

**Comment by kekekeks** [comment]
After taking a look at heuristics in your repo, I'm not sure we can use those. We might just follow [what Qt6 does](https://doc-snapshots.qt.io/qt6-dev/highdpi.html#configuring-x11), which is just reading `Xft.dpi` from Xresources with an option to enable EDID-based detection via environment variable.

**Comment by HermanKirshin** [comment]
That would be good, because 
Xft.dpi will respect system settings perfectly on KDE (in any configuration - pure X11, Wayland with X11 scaled by system, Wayland with X11 scaled by themselves) and Cinnamon (both with and without fractional scaling enabled), both in multi-monitor scenarios with different settings. These two environments today have the best scaling support out of the box.

It will work on Fedora with Gnome+Wayland but unlike KDE will not respect mixed monitor settings, and still will not work in Ubuntu without messing with configs because it does not change Xft.dpi on settings change.

It will not respect system settings on MATE, XFCE, LXQt and some others, but I think they are not frequent choice for HiDpi configurations.

It will work same way both on real and virtual machines.

So I think reading Xft.dpi is way more better approach than current magic with EDID, because unlike EDID it at least works and moreover - it will handle correctly at least half of cases.

**Comment by HermanKirshin** [comment]
And why trying to also read values from libgtk is not an option?

**Comment by kekekeks** [comment]
Attempting to use GTK from the UI thread will break our native dialogs since they require a separate event loop on a background thread.

**Comment by HermanKirshin** [comment]
But calling it using GtkInteropHelper.RunOnGLibThread with preceding XInitThreads works fine and does not break native dialogs

**Comment by Lordfirespeed** [comment]
Any update on this ?

**Comment by anyoneyun** [comment]
Any updates?

**Comment by kekekeks** [comment]
Avalonia [currently reads Xft.dpi value from XRDB](https://github.com/AvaloniaUI/Avalonia/blob/7a7bee4cb4fa054d6780acb3b9138b822c3152dc/src/Avalonia.X11/Screens/X11Screens.Scaling.cs#L58-L61), which is the only thing we can do since WM/DE authors **_simply refuse_** to provide accurate per-monitor scaling information for X11 apps.

Avalonia also respects QT environment variables (QT_SCREEN_SCALE_FACTOR/QT_AUTO_SCREEN_SCALE_FACTOR) for configuring per-monitor scaling, so if you set such envs in /etc/profile those should work.

**Comment by l0b0** [comment]
> Avalonia [currently reads Xft.dpi value from XRDB](https://github.com/AvaloniaUI/Avalonia/blob/7a7bee4cb4fa054d6780acb3b9138b822c3152dc/src/Avalonia.X11/Screens/X11Screens.Scaling.cs#L58-L61), which is the only thing we can do since WM/DE authors **_simply refuse_** to provide accurate per-monitor scaling information for X11 apps.
> 
> Avalonia also respects QT environment variables (QT_SCREEN_SCALE_FACTOR/QT_AUTO_SCREEN_SCALE_FACTOR) for configuring per-monitor scaling, so if you set such envs in /etc/profile those should work.

Is this something which needs to be programmed into each application? I've tried a few things to [get Scarab to scale](https://github.com/fifty-six/Scarab/issues/109), but none of these seem to work:

- `xrdb -query | rg '^Xft.dpi'` reports 192, but `Scarab` is not scaled.
- `QT_SCREEN_SCALE_FACTOR=2 Scarab` does not change anything.
- `QT_AUTO_SCREEN_SCALE_FACTOR=1 Scarab` does not change anything.

**Comment by teohhanhui** [comment]
@l0b0 Have you tried:

```
XFT_DPI=$(xrdb -get Xft.dpi)
AVALONIA_GLOBAL_SCALE_FACTOR=$(echo "scale=2; $XFT_DPI/96" | bc)
export AVALONIA_GLOBAL_SCALE_FACTOR
```

**Comment by ShalokShalom** [comment]
It worked for me with `AVALONIA_GLOBAL_SCALE_FACTOR=2 lumafly`

**Comment by l0b0** [comment]
Has Avalonia scaling changed since the above? Because I can no longer scale Scarab by setting `AVALONIA_GLOBAL_SCALE_FACTOR`
